File: android/app/src/main/kotlin/com/hellenicdir/ui/auth/AuthViewModel.kt
package com.hellenicdir.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hellenicdir.core.auth.EncryptedTokenStorage
import com.hellenicdir.core.network.ApiService
import com.hellenicdir.data.remote.dto.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Current version of the Privacy Policy — must match what is displayed in the UI. */
private const val PRIVACY_POLICY_VERSION = "2025-01-01"

data class AuthUiState(
    val user: UserDto? = null,
    val isLoading: Boolean = true,
    val error: String? = null,
)

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val api: ApiService,
    private val tokenDataStore: EncryptedTokenStorage,
) : ViewModel() {

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch { restoreSession() }
    }

    // MARK: Session restoration

    private suspend fun restoreSession() {
        val token = tokenDataStore.getAccessToken()
        if (token == null) {
            _uiState.update { it.copy(isLoading = false) }
            return
        }
        try {
            val response = api.getMe()
            _uiState.update { it.copy(user = response.user, isLoading = false) }
        } catch (e: Exception) {
            tokenDataStore.clearTokens()
            _uiState.update { it.copy(isLoading = false) }
        }
    }

    // MARK: Login

    fun login(email: String, password: String, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val response = api.login(LoginRequest(email.trim().lowercase(), password))
                tokenDataStore.saveTokens(response.accessToken, response.refreshToken)
                _uiState.update { it.copy(user = response.user, isLoading = false) }
                onSuccess()
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message ?: "Login failed") }
            }
        }
    }

    // MARK: Registration

    /**
     * Registers a new account.
     * Consent parameters must be `true` — the UI enforces checkbox ticking before
     * calling this function. The consent version and timestamps are recorded
     * server-side for CCPA/PIPEDA compliance audit trails.
     */
    fun register(
        fullName: String,
        email: String,
        password: String,
        phone: String?,
        onSuccess: () -> Unit,
    ) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                api.register(
                    RegisterRequest(
                        email = email.trim().lowercase(),
                        password = password,
                        fullName = fullName.trim(),
                        phone = phone?.ifBlank { null },
                        privacyPolicyVersion = PRIVACY_POLICY_VERSION,
                        consentPrivacyPolicy = true,
                        consentTerms = true,
                        consentSensitiveData = true,
                    )
                )
                _uiState.update { it.copy(isLoading = false) }
                onSuccess()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, error = e.message ?: "Registration failed")
                }
            }
        }
    }

    // MARK: Logout

    fun logout(onComplete: () -> Unit) {
        viewModelScope.launch {
            val refreshToken = tokenDataStore.getRefreshToken()
            try { api.logout(LogoutRequest(refreshToken)) } catch (_: Exception) {}
            tokenDataStore.clearTokens()
            _uiState.update { it.copy(user = null) }
            onComplete()
        }
    }

    // MARK: Account deletion (GDPR/CCPA right to erasure)

    /**
     * Requests account deletion. The backend soft-deletes the account and revokes
     * all tokens; hard-deletion is scheduled after the 30-day regulatory grace period.
     *
     * On success, clears local tokens and resets UI state. On failure, surfaces the
     * error message so the UI can prompt the user to contact support.
     */
    fun deleteAccount(onComplete: () -> Unit, onError: (String) -> Unit) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                api.deleteAccount()
                tokenDataStore.clearTokens()
                _uiState.update { it.copy(user = null, isLoading = false) }
                onComplete()
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false) }
                onError(e.message ?: "Account deletion failed. Please contact support.")
            }
        }
    }
}

File: ios/HellenicDirectory/Features/Auth/ViewModels/AuthViewModel.swift
import Foundation
import UIKit
import Combine
import OSLog

/// Current Privacy Policy version — must match the version displayed in the registration UI.
private let PRIVACY_POLICY_VERSION = "2025-01-01"

private let logger = Logger(subsystem: "com.hellenicdir", category: "AuthViewModel")

// MARK: - Response / request types

struct UserResponse: Decodable {
    let user: AppUser
}

struct AppUser: Decodable, Identifiable {
    let id: String
    let email: String
    let fullName: String
    let phone: String?
    let avatarUrl: String?
    let appRole: String
}

struct LoginRequest: Encodable {
    let email: String
    let password: String
    let deviceId: String?
}

struct LoginResponse: Decodable {
    let user: AppUser
    let accessToken: String
    let refreshToken: String
}

/// Registration request body.
/// Consent fields are required by the backend per CCPA/PIPEDA sensitive-data rules.
struct RegisterRequest: Encodable {
    let email: String
    let password: String
    let fullName: String
    let phone: String?
    /// Version of the Privacy Policy the user accepted (e.g. "2024-01-01").
    let privacyPolicyVersion: String
    let consentPrivacyPolicy: Bool
    let consentTerms: Bool
    let consentSensitiveData: Bool
}

// MARK: - AuthViewModel

@MainActor
class AuthViewModel: ObservableObject {
    @Published var currentUser: AppUser?
    @Published var isLoading = true
    @Published var error: String?

    var isAdmin: Bool { currentUser?.appRole == "OWNER" || currentUser?.appRole == "ADMIN" }
    var isOwner: Bool { currentUser?.appRole == "OWNER" }

    init() {
        Task { await restoreSession() }
    }

    /// Attempts to restore the authenticated session from the stored Keychain tokens.
    func restoreSession() async {
        let hasToken = await KeychainManager.shared.accessToken != nil
        guard hasToken else {
            isLoading = false
            return
        }
        do {
            let response: UserResponse = try await APIClient.shared.get("/auth/me")
            currentUser = response.user
        } catch {
            logger.warning("Session restore failed — clearing tokens: \(error.localizedDescription, privacy: .public)")
            await KeychainManager.shared.clearAll()
        }
        isLoading = false
    }

    /// Authenticates with email and password and persists the issued tokens to Keychain.
    func login(email: String, password: String) async throws {
        // UIDevice.identifierForVendor must be accessed on the main thread.
        // This method is @MainActor so the call is safe here.
        let deviceId = UIDevice.current.identifierForVendor?.uuidString

        let response: LoginResponse = try await APIClient.shared.post(
            "/auth/login",
            body: LoginRequest(email: email, password: password, deviceId: deviceId)
        )
        await KeychainManager.shared.setAccessToken(response.accessToken)
        await KeychainManager.shared.setRefreshToken(response.refreshToken)
        currentUser = response.user
    }

    /// Registers a new account.
    ///
    /// Consent acknowledgements are passed as `true` because the registration UI
    /// requires all three checkboxes to be ticked before this function is callable.
    ///
    /// - Parameters:
    ///   - privacyPolicyVersion: The version string of the Privacy Policy the user accepted.
    func register(
        fullName: String,
        email: String,
        password: String,
        phone: String?
    ) async throws {
        let _: UserResponse = try await APIClient.shared.post(
            "/auth/register",
            body: RegisterRequest(
                email: email,
                password: password,
                fullName: fullName,
                phone: phone,
                privacyPolicyVersion: PRIVACY_POLICY_VERSION,
                consentPrivacyPolicy: true,
                consentTerms: true,
                consentSensitiveData: true
            )
        )
    }

    /// Signs out by revoking the refresh token and clearing Keychain credentials.
    func logout() async {
        let refreshToken = await KeychainManager.shared.refreshToken
        struct LogoutBody: Encodable { let refreshToken: String? }
        try? await APIClient.shared.post("/auth/logout", body: LogoutBody(refreshToken: refreshToken)) as EmptyResponse
        await KeychainManager.shared.clearAll()
        currentUser = nil
    }

    /// Requests account deletion (GDPR/CCPA right to erasure).
    /// Revokes all tokens server-side and clears local credentials.
    func deleteAccount() async throws {
        try await APIClient.shared.delete("/auth/account")
        await KeychainManager.shared.clearAll()
        currentUser = nil
    }
}
